packages/rs-drive-abci/src/abci/handlers.rs

```diff
@@ -32,720 +32,1052 @@
-                                block_height, day
-                            )
-                        });
-
-                    // Should pay to all proposers for epoch 0, when epochs 1 started
-                    if epoch_index != 0 && epoch_change {
-                        assert!(block_end_response.proposers_paid_count.is_some());
-                        assert!(block_end_response.paid_epoch_index.is_some());
-
-                        assert_eq!(
-                            block_end_response.proposers_paid_count.unwrap(),
-                            proposers_count
-                        );
-                        assert_eq!(block_end_response.paid_epoch_index.unwrap(), 0);
-                    } else {
-                        assert!(block_end_response.proposers_paid_count.is_none());
-                        assert!(block_end_response.paid_epoch_index.is_none());
-                    };
-                }
-            }
+            Ok(proto::ResponseVerifyVoteExtension {
+                status: VerifyStatus::Reject.into(),
+            })
         }
+    }
 
-        #[test]
-        fn test_chain_halt_for_36_days() {
-            // TODO refactor to remove code duplication
-
-            let mut platform = setup_platform_raw(Some(PlatformConfig {
-                verify_sum_trees: false,
-                ..Default::default()
-            }));
-
-            let mut core_rpc_mock = MockCoreRPCLike::new();
-
-            core_rpc_mock
-                .expect_get_block_hash()
-                // .times(1) // TODO: investigate why it always n + 1
-                .returning(|_| {
-                    Ok(BlockHash::from_hex(
-                        "0000000000000000000000000000000000000000000000000000000000000000",
-                    )
-                    .unwrap())
-                });
-
-            core_rpc_mock
-                .expect_get_block_json()
-                // .times(1) // TODO: investigate why it always n + 1
-                .returning(|_| Ok(json!({})));
-
-            platform.core_rpc = Box::new(core_rpc_mock);
-
-            let transaction = platform.drive.grove.start_transaction();
-
-            // init chain
-            let init_chain_request = static_init_chain_request();
-
-            platform
-                .init_chain(init_chain_request, Some(&transaction))
-                .expect("should init chain");
-
-            // setup the contract
-            let contract = platform.create_mn_shares_contract(Some(&transaction));
-
-            let genesis_time = Utc::now();
+    fn finalize_block(
+        &self,
+        request: RequestFinalizeBlock,
+    ) -> Result<ResponseFinalizeBlock, ResponseException> {
+        let transaction_guard = self.transaction.read().unwrap();
 
-            let epoch_2_start_day = 37;
+        let transaction = transaction_guard.as_ref().ok_or(Error::Execution(
+            ExecutionError::NotInTransaction(
```

Comment:
Is there a possibility that the transaction won't be available here? And if so, can it be non-deterministic behavior? I guess not right now since we don't have any concurrency, but potentially can be in the future

Reply:
Everything is fine here. The transaction won't be here only if we have a bug in prepare/process proposal.